clients/src/main/java/org/apache/kafka/clients/FetchSessionHandler.java

```diff
@@ -0,0 +1,340 @@
+            response.responseData().keySet());
+        if (!missing.isEmpty()) {
+            log.warn("Session handler for node {} received a full fetch response " +
+                "that did not contain expected partition(s) {}", node, missing);
+            return false;
+        }
+        return true;
+    }
+
+    /**
+     * Verify that the partitions in an incremental fetch response are contained in the session.
+     *
+     * @param response  The response.
+     * @return          True if the incremental fetch response partitions are valid.
+     */
+    private boolean verifyIncrementalFetchResponseParts(FetchResponse response) {
+        String missing = findMissing(MAX_MISSING_TO_LOG, response.responseData().keySet(),
+            sessionParts.keySet());
+        if (!missing.isEmpty()) {
+            log.warn("Session handler for node {} received an incremental fetch response " +
+                "that contained unexpected partition(s) {}", node, missing);
+            return false;
+        }
+        return true;
+    }
+
+    public boolean handleResponse(FetchResponse response) {
+        if (response.error() == Errors.NONE) {
+            log.debug("Session handler for node {} received FetchResponse for {} with session " +
+                    "id {} and {} partition(s).", node, nextMetadata,
+                response.sessionId(), response.responseData().size());
+            if (nextMetadata.isFull()) {
+                if (!verifyFullFetchResponseParts(response)) {
+                    nextMetadata = FetchMetadata.INITIAL;
+                    return false;
+                } else {
+                    if (response.sessionId() == INVALID_SESSION_ID) {
+                        nextMetadata = FetchMetadata.INITIAL;
+                    } else {
+                        nextMetadata = FetchMetadata.newIncremental(response.sessionId());
+                    }
+                    return true;
+                }
+            } else {
+                if (!verifyIncrementalFetchResponseParts(response)) {
+                    nextMetadata = nextMetadata.nextCloseExisting();
+                    return false;
+                } else {
+                    nextMetadata = nextMetadata.nextIncremental();
+                    return true;
+                }
+            }
+        } else if (response.error() == Errors.FETCH_SESSION_ID_NOT_FOUND) {
+            log.info("Session handler for node {} received FETCH_SESSION_ID_NOT_FOUND error on {}.",
+                node, nextMetadata);
+            nextMetadata = FetchMetadata.INITIAL;
+            return false;
+        } else {
+            log.info("Session handler for node {} received response error {} for {}.",
+                node, response.error(), nextMetadata);
+            nextMetadata = nextMetadata.nextCloseExisting();
+            return false;
+        }
+    }
+
+    public void handleError(Throwable t) {
```

Comment:
Can you document when this should be used? From what I can see, it only has one use in the replica fetcher at the moment.

Reply:
> Can you document when this should be used?

It should be called whenever we couldn't communicate with the broker.  Basically it clears the fetch session state.

> From what I can see, it only has one use in the replica fetcher at the moment.

Good point-- it should have been called from `Fetcher.java` as well.  Fixed.

Reply:
I also added some JavaDoc